[&str] update about.md (#1227)
exercism/rust: concepts/&str/about.md

# About

> There are two ways that Rust implements an array of characters: [str][str] and [String][String].

Reviewer: ```suggestion
There are two ways that Rust implements an array of characters: [`str`][str] and [`String`][String].
```

It seems like we are establishing a convention of using code formatting when discussing types. Sound fair?

Author: I had it that way at first, but then noticed that it hid the link unless you mouse over it, so I removed the back ticks. But I don't feel strongly about it if you want me to put them back in.

Reviewer 2: I prefer to render types with code formatting. If exercism's stylesheet hides links in code formatting, that's a stylesheet bug; we should fix the stylesheet instead of leaving out semantic markers to work around it.

`str`, also known as a "string slice", is usually used with a reference as `&str`.

To know why, it may help to understand why the following line of code will not compile

```rust

> const my_const: str = "Hello, World!";

Reviewer 2: ```suggestion
const my_const: str = *"Hello, World!";
```

We'd need the dereference there to avoid a type error; string literals have type `&'static str`. For example, this is legal:

```rust
const HELLO: [u8; 12] = *b"Hello, world";
```

Author: The suggested line

```rust
const my_const: str = *"Hello, World!";
```
results in the same error. We need a reference, not a dereference.

Author: Yes, I know what you mean that the dereference is meant to bring the  `&'static str` to a str, but I'm not sure it might be confusing to throw the deref operator in there for students who need help with `&str`. I wanted to avoid getting into derefs in order to keep the focus on `&str`, but I don't feel strongly about putting the `*` in there if you think it won't confuse the matter for newer students.

Author: I was tempted to cover `&'static str` but decided to keep lifetimes out of it.

Author: I get the type mismatch error on a let, but not a const. Perhaps it should have the deref operator in case the compiler ever changes to squawk  about the mismatch. Sorry for all the thinking out loud...

// the size for values of type `str` cannot be known at compilation time
// the trait `Sized` is not implemented for `str`
```

Even though the literal "Hello, World!" length can be known at compilation time, as a `str` it is a sequence of unknown size.
But as a _reference_ to a literal (`&str`), it has a "fat pointer" which holds the address and length of the literal in memory.
Fat pointers are of a known size, and so this line will compile

```rust
const my_const: &str = "Hello, World!";
```

> A mutable string slice (`mut &str`) can have its valued replaced, but not modified in place.

Reviewer 2: ```suggestion
A mutable binding to a string slice can have its valued replaced, but not modified in place.
```

Two changes in the suggestion:

- Add the word "binding", because that's what's mutable, not the reference
- remove the type, because `mut &str` isn't really a thing you'll see in rust.

Example

```rust
let mut my_str = "Hello, World!";
my_str = "Hi!";
let mut my_string: String = String::from("Hello");
my_string.push_str(", World!"); // there is no equivalent function to modify &str
```

It is less constraining to use `&str` instead of `&String` when defining a parameter for a function, as this will not compile

```rust
pub fn main() {
    let my_str: &str = "Hello, World!";
    say_hi(my_str);
}

fn say_hi(phrase: &String) {
    println!("{:?}", phrase);
}

// mismatched types
// expected reference `&String`
// found reference `&str`
```

A `&str` will not be [coerced][coercion] into a `&String`. But this will compile

```rust
pub fn main() {

>     let my_string: String = String::from("Hello, World!");

Reviewer 2: ```suggestion
    let my_string = String::from("Hello, World!");
```

An explicit type there is redundant.

Author: Yeah, I had it explicit for pedagogic purposes, but I think you're right to keep it best practice.
